### ISSUE TITLE
How supported do we consider non-keyextract users?

### ISSUE THREAD

**Issue by Manishearth** [unicode-org/icu4x#2975]
This discussion keeps cropping up and I feel like we should properly discuss it in its full context as opposed to through the lens of different issues. (In particular, I think Shane and I have diametrically opposed opinions on this)

Currently, ICU4X datagen supports "key extraction", where given a binary it automatically figures out the keys that are involved and generate data for those. It also supports manually providing these keys.

I believe everyone on the team is on board with "keyextract" being the recommended tool. However, there seems to be disagreement on _how_ likely we consider it that someone will be off the happy path, how much help we wish to provide there, and how "first-class" that scenario is considered.

Some questions I would like to get a clear answer on:

 - Do we attempt to make non-keyextract user flows as well-supported as those with keyextract (with the equivalent amount of documentation, etc)? https://github.com/unicode-org/icu4x/issues/2685 is a part of this
 - If a user flow is not supported by keyextract, is this considered a bug, or a nice-to-have?
 - If a feature (for example, [baked data in FFI](https://github.com/unicode-org/icu4x/pull/2947)) is not supported by keyextract, is this considered a blocker for the feature, or nice-to-have?

(I'll post my own opinions in a comment so as not to pollute the main issue body)

cc @sffc @zbraniecki @robertbastian @nordzilla

**Comment by Manishearth** [comment]
TLDR: I do think we should consider this a first-class supported use case, but strongly discouraged in the docs. If workflows or features don't work with keyextract, that is not necessarily a blocker, but it is a tradeoff to be considered among others.

Overall I think that while keyextract is convenient, it is a rather nontraditional workflow and a lot of users are going to have problems with it, and we should not make life harder for them.

At the very least, we ought to make sure keyextract can be run in two phases: one where a key file is generated, and then one where datagen actually generates the data, since this is more friendly for build systems that can cache and skip steps.

### The build system issue

With that aside, I think the fact that keyextract depends on the final binary is going to be an issue for a lot of build systems.


We already see this with databake, where even Cargo doesn't really have the ability to handle this: the best workflow we have here is that you write your code to use a dummy provider, run datagen on that, and edit it to use the real baked data. We can probably make this easier by providing a well-documented workflow where you dump an everything-baked provider in the bake directory first and then overwrite it later.

But that's not going to cut it for e.g. Bazel, where generated code is expected to be handled via genrule, and cyclic dependencies are really not a thing. (In general, build systems strongly avoid trying to deal with cyclic dependencies — Cargo _seems_ to allow cycles via dev deps but it [doesn't actually at the fundamental level](https://docs.rs/guppy/latest/guppy/graph/struct.Cycles.html#why-cargo-dependency-graphs-have-cycles) )

_Technically_ we could do something really complicated where the databake genrule can mess with its dependency spec and create shadow dependencies to do the two-step build. This would almost certainly be considered a major hack and very brittle.


### Datagen output as a derived artifact

I also don't think this problem is unique to databake: it's a more stark member of a class of problems that basically boils down to "wanting the output of datagen to not change unexpectedly", where with datagen it's the build system doing the expecting.

But I can imagine other users wanting to closely control data being shipped: they'd want to notice when the keys change, perhaps due to an errant API call. One way to do that is to golden-test a keyextracted keyfile and check that in CI, but I do think that assumes a bunch about a particular project's CI workflows. I do think it should be valid for a keyfile to be the primary source of truth, and for users to have to update it explicitly.

Risk of the keyfile going out of date can still be mitigated with various kinds of tests using keyextract if necessary.


Fundamentally I think the disagreement here is whether we consider the list of keys to be a pure derived property or a separate thing, and I personally feel strongly that "what data we want to use" is a separate thing that a project may review separately.


### Things like AnyCalendar

Finally, we have things like AnyCalendar: where there _is_ a typed solution, but there is also a type-erased one that will end up pulling in all of the data. We may end up needing similar stuff for unit formatting and currency formatting.


Keyextract doesn't cut it here if the user wants finer-grained control; though I can see a potential future where these specific things can be configured via `datagen`'s config.

**Comment by sffc** [comment]
There are two related but separate mechanisms for data slicing:

1. Automatic DCE of BakedProvider
2. Keyextract via Datagen

The "Automatic DCE" data slicing is the easiest and most foolproof method, and we should favor it whenever possible. It works so long as we avoid converting the BakedProvider to an AnyProvider. I guess what I'm saying is that when we _can_, we _should_ try to design the data provider systems such that if BakedProvider is being used, then we get Automatic DCE data slicing.

Keyextract is most useful when Postcard is being used.

I think it's totally fine if Datagen runs on an explicit list of keys, with the consistency between the list and Keyextract being a unit test. In fact, I consider this the more robust code path since it makes the intermediate state more explicit.

**Comment by sffc** [comment]
Also, to respond to @Manishearth's comment in https://github.com/unicode-org/icu4x/pull/2947#discussion_r1066525051

> that's going to become a problem again if combinators are used

We have a big suite of tools for users who want fancy combinatorial pluggable dynamic data pipelines, Keyextract playing a central role. However, BakedProvider's primary use case are clients who don't want all that pizzazz and are okay shipping their app with all the data it needs. As long as we support a "bare" BakedProvider and one with a single vertical fallback combinator, we reach most of these users. If we can support these users with automatic DCE data slicing, then we should.

The main use case I can see for putting BakedProvider in a combinator is if you want to combine it with some sort of dynamic data loading, such as downloading more locales on demand. This is a problem that is not yet fully solved due to the registry issue. I'd like to see if the solution could retain our ability to use automatic data slicing.

**Comment by Manishearth** [comment]
> However, BakedProvider's primary use case are clients who don't want all that pizzazz and are okay shipping their app with all the data it needs

I think it's more accurate to say that it's targeted for people who want to ship their app with _some_ of the data needed. It was designed in response to Firefox having exactly this need: not wanting to completely swap out the data loading mechanisms but rather wanting to be able to mix and match.



-----

Also, the DCE thing clashes with keeping dependencies low. rustc wants to use baked list format, but it's not using the metacrate; so they specify a list of keys.

I think my core contention here is that keyextract is super convenient, but it can _very easily_ clash with others' constraints, and we should treat it as preferred but not the only first-class way of doing things.

**Comment by sffc** [comment]
I think solving https://github.com/unicode-org/icu4x/issues/2945 will solve a lot of the issues in this thread.

**Comment by Manishearth** [comment]
I think it solves it for the casual users of the library, but not more complex users, and  my point about others' constraints is really only an issue for complex users in the first place.

**Comment by sffc** [comment]
Simply put, advanced users should use keyextract, and casual users should use automatic DCE, and we should do as much as we can to get as many users into the "casual" category as possible, while still considering keyextract as a first class feature for power users.

**Comment by Manishearth** [comment]
Yes, absolutely, but I think my point is that there are many reasons for which advanced users will not be able to use keyextract because keyextract imposes a lot of weird constraints; and we should also consider non-keyextract users to be first class, if discouraged by default.

**Comment by sffc** [comment]
I'm not sure I follow. If a keyfile is being used, _that is still a keyextract workflow_, because a keyfile needed to be generated somehow, and it should be kept in sync with a test that runs keyextract.

There are cases where datagen may accept additional flags to tailor the data for collation, anycalendar, segmenter, etc., but this is orthogonal to the question of the set of keys in code.

**Comment by Manishearth** [comment]
I think I don't quite agree with that ontology of it still being a "keyextract workflow" because I'm talking about whether or not the humans need to care about the actual keys: and they might even if they have a test.

Also, this is rather normative:

> and it should be kept in sync with a test that runs keyextract


Yes, if someone is using a keyfile, they _probably_ should be verifying that it is correct. However, there are many ways in which they may choose to do that, and many workflows involving far more human steps there. There _will_ be cases where they'll need to care about the actual keys, and my hope is that they will be well-supported there.

Here's another example: there's a codebase filled with libraries that turns into a large binary at the end. Individual libraries are maintained by different people, and they wish to each do their own data generation. Now our response to that may be "don't do that, make data globally", but that may not be an option for them, especially if they want some level of modularity. In general I'm very wary of situations where our answer to a large codebase is "don't do that", because often these things are borne out of the interactions of complex constraints.

This has been my repeated experience with large codebases: they have a lot of these constraints, and anything _weird_ is something they may choose to opt out of. I feel like keyextract workflows fall under this category.

------

What I consider a "non keyextract user flow" is one where the user either (a) chooses not to use keyextract for workflow reasons, or (b) disagrees with the output of keyextract for their use case.

The argument I am attempting to make here is that while we may wish to consider (b) a bug in datagen's configurability, or consider (a) to be a bug in keyextract, I'm not sure if that's actually feasable for _all_ cases there, and we should try to pave all paths, and be _okay_ with cases where keyextract won't work due to valid tradeoffs, rather than trying to fit it into everything.

**Comment by Manishearth** [comment]
Discussed further with @sffc 


To the following point:

> Do we attempt to make non-keyextract user flows as well-supported as those with keyextract (with the equivalent amount of documentation, etc)?

we think the answer should be "yes", with caveats.

The caveats are that we should make sure it is easy and documented for having such workflows, however:

 - We prominently document keyextract as the recommended workflow
 - We note that "if you can't rely on this, please see [docs for non keyextract workflows]", and specify that we recommend relying on keyextract as much as possible
 - We document known reasons for needing a non keyextract workflow (one scenario I brought up was data splitting). We also document whether we have plans to make those work with keyextract eventually, so users can subscribe to relevant issues if necessary.
 - We ask people to let us know their reasons if they are not amongst the list of documented reasons

For these two points:

> - If a user flow is not supported by keyextract, is this considered a bug, or a nice-to-have?
>
> - If a feature (for example, https://github.com/unicode-org/icu4x/pull/2947) is not supported by keyextract, is this considered a blocker for the feature, or nice-to-have?

we consider it a bug and a stabilization blocker, though whilst building a feature it's not necessary to immediately figure this out. There may be some exceptions.


Currently calendar filtering is one such use case where keyextract isn't sufficient, however we support typed and Any calendars and that ought to cover _most_ use cases; we can get people to ask us if they have their own.

-----


Also, the general picture @sffc drew was that he considers any workflow that involves deciding what goes in a keyfile manually to be a hack. (assuming that "keyfile was written manually but is tested by keyextract" doesn't count as actually being manual, since you're not _deciding_ what goes in there by hand).

 It is fine if you have a keyfile that you split up by hand, or if you merge keyfiles by hand, but that is also something that can be tested in an automated fashion.

I think this is probably fine.

**Comment by sffc** [comment]
> We note that "if you can't rely on this, please see [docs for non keyextract workflows]", and specify that we recommend relying on keyextract as much as possible

I'd like to see two sets of docs

1. Recommended datagen workflow tutorials
2. Comprehensive technological explainer on how the data pipeline works

> We document known reasons for needing a non keyextract workflow (one scenario I brought up was data splitting).

As you noted at the bottom of your comment, I don't actually see these as "non keyextract workflows", and I also therefore don't see the need for a different docs page for them.

**Comment by Manishearth** [comment]
> and I also therefore don't see the need for a different docs page for them.

I think we still need to document how keys and stuff work since they will need to know about them to split data sensibly.
